service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/azure/container/CopyAzureStorageContainerDefinitionStep.java

```diff
@@ -0,0 +1,155 @@
+      ResourceDao resourceDao,
+      LandingZoneApiDispatch landingZoneApiDispatch,
+      ControlledAzureStorageContainerResource sourceContainer,
+      ControlledResourceService controlledResourceService,
+      CloningInstructions resolvedCloningInstructions) {
+    this.userRequest = userRequest;
+    this.sourceContainer = sourceContainer;
+    this.controlledResourceService = controlledResourceService;
+    this.resolvedCloningInstructions = resolvedCloningInstructions;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    var inputParameters = flightContext.getInputParameters();
+    var workingMap = flightContext.getWorkingMap();
+    validateInputs(inputParameters, workingMap);
+
+    // get the inputs from the flight context
+    var destinationResourceName =
+        FlightUtils.getInputParameterOrWorkingValue(
+            flightContext,
+            WorkspaceFlightMapKeys.ResourceKeys.RESOURCE_NAME,
+            WorkspaceFlightMapKeys.ResourceKeys.PREVIOUS_RESOURCE_NAME,
+            String.class);
+    var description =
+        FlightUtils.getInputParameterOrWorkingValue(
+            flightContext,
+            WorkspaceFlightMapKeys.ResourceKeys.RESOURCE_DESCRIPTION,
+            WorkspaceFlightMapKeys.ResourceKeys.PREVIOUS_RESOURCE_DESCRIPTION,
+            String.class);
+    var destinationWorkspaceId =
+        inputParameters.get(
+            WorkspaceFlightMapKeys.ControlledResourceKeys.DESTINATION_WORKSPACE_ID, UUID.class);
+    var destinationContainerName =
+        inputParameters.get(
+            WorkspaceFlightMapKeys.ControlledResourceKeys.DESTINATION_CONTAINER_NAME, String.class);
+    var destinationResourceId =
+        inputParameters.get(
+            WorkspaceFlightMapKeys.ControlledResourceKeys.DESTINATION_RESOURCE_ID, UUID.class);
+    var destStorageAccountId =
+        flightContext
+            .getWorkingMap()
+            .get(
+                WorkspaceFlightMapKeys.ControlledResourceKeys
+                    .DESTINATION_STORAGE_ACCOUNT_RESOURCE_ID,
+                UUID.class);
+
+    ControlledAzureStorageContainerResource destinationContainerResource =
+        buildDestinationControlledAzureContainer(
+            sourceContainer,
+            destStorageAccountId,
+            destinationWorkspaceId,
+            destinationResourceId,
+            destinationResourceName,
+            description,
+            destinationContainerName);
+    ApiAzureStorageContainerCreationParameters destinationCreationParameters =
+        new ApiAzureStorageContainerCreationParameters()
+            .storageContainerName(destinationContainerName)
+            .storageAccountId(destStorageAccountId);
+    ControlledResourceIamRole iamRole =
+        IamRoleUtils.getIamRoleForAccessScope(sourceContainer.getAccessScope());
+
+    // create the container
+    ControlledAzureStorageContainerResource clonedContainer =
```

Comment:
What happens if this is the second execution of this step? E.g., the pod is restarted on an upgrade.

Reply:
A good question, the name will collide when WSM goes to create the resource in the DB and we'll get a DuplicateResourceException. I'm not sure how we can reliably determine if the duplicate is due to retry vs a genuine naming collision at this point in the flight -- looking around a bit it seems like we are susceptible to this same issue in our other cloning flights as well (i.e, this [step](https://github.com/DataBiosphere/terra-workspace-manager/blob/7f51c97637563277723d882a256ada141f2e85e8/service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/bucket/CopyGcsBucketDefinitionStep.java#L119) for gcs buckets and this [step](https://github.com/DataBiosphere/terra-workspace-manager/blob/7f51c97637563277723d882a256ada141f2e85e8/service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/dataset/CopyBigQueryDatasetDefinitionStep.java#L111-L115) for bq datasets).

It may make sense to handle that exception here and move on to blob cloning -- thoughts @ddietterich ?

Reply:
I think I don't have the whole picture in my head. If you can assure uniqueness of the resource by inspecting the WSM metadata, then you don't have a problem here. I'm not sure how you do that, but if all containers are allocated through WSM and you scan across all workspaces, it seems possible. If that is so, you'd just need to tolerate the duplicate error because you could be sure ahead of time that the name was unique and this flight was responsible for creating it.

Sometimes it is not possible to completely eliminate the window. One technique to reduce the window is to have an existence check step ahead of the create step. If the existence check step fails with a duplicate, we fail the flight with "duplicate". If the create step fails with a duplicate, we know with high probability that it was us retrying and we continue.

The bad case is when the duplicate entry exists before this flight, the create fails with duplicate, and then the create-undo deletes the object during cleanup. We just deleted someone else's cloud resource!

Reply:
If we check for the resource name within the workspace prior to the creation step, I think we should be able to reliably retry the copy definition step. If the resource already exists prior to the start of this flight (i.e,. someone else's container), the flight will fail out prior to the copy definition step. I'll add a step to do this.

Reply:
I've added a check for duplicate step [here](https://github.com/DataBiosphere/terra-workspace-manager/blob/1b725dff35e3e1667f36eb1e7aad95f19269907f/service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/azure/container/CloneControlledAzureStorageContainerResourceFlight.java#L102) and now catch the duplicate resource exception [here](https://github.com/DataBiosphere/terra-workspace-manager/blob/1b725dff35e3e1667f36eb1e7aad95f19269907f/service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/azure/container/CopyAzureStorageContainerDefinitionStep.java#L112-L123)